subsys/bluetooth/services/bas.c

```diff
@@ -1,73 +1,129 @@
+static struct bt_gatt_ccc_cfg charge_lvl_ccc_cfg[BT_GATT_CCC_MAX];
 
-static struct bt_gatt_ccc_cfg  blvl_ccc_cfg[BT_GATT_CCC_MAX] = {};
-static u8_t simulate_blvl;
-static u8_t battery = 100;
+static struct device *battery;
+static struct sensor_value charge_lvl;
+static bool notif_enabled;
 
-static void blvl_ccc_cfg_changed(const struct bt_gatt_attr *attr,
-				 u16_t value)
+int bas_notify(void);
+
+static void charge_lvl_ccc_cfg_changed(const struct bt_gatt_attr *attr,
+				       u16_t value)
 {
-	simulate_blvl = (value == BT_GATT_CCC_NOTIFY) ? 1 : 0;
+	ARG_UNUSED(attr);
+
+	notif_enabled = (value == BT_GATT_CCC_NOTIFY);
 }
 
-static ssize_t read_blvl(struct bt_conn *conn, const struct bt_gatt_attr *attr,
-			 void *buf, u16_t len, u16_t offset)
+static void charge_lvl_fetch(void)
 {
-	const char *value = attr->user_data;
+	int rc;
 
-	return bt_gatt_attr_read(conn, attr, buf, len, offset, value,
-				 sizeof(*value));
-}
+	__ASSERT(battery != NULL, "No binding to battery sensor!\n");
 
-/* Battery Service Declaration */
-static struct bt_gatt_attr attrs[] = {
-	BT_GATT_PRIMARY_SERVICE(BT_UUID_BAS),
-	BT_GATT_CHARACTERISTIC(BT_UUID_BAS_BATTERY_LEVEL,
-			       BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY,
-			       BT_GATT_PERM_READ, read_blvl, NULL, &battery),
-	BT_GATT_CCC(blvl_ccc_cfg, blvl_ccc_cfg_changed),
-};
+	rc = sensor_sample_fetch_chan(battery, SENSOR_CHAN_SOC);
+	if (rc) {
+		return;
+	}
 
-static struct bt_gatt_service bas_svc = BT_GATT_SERVICE(attrs);
+	rc = sensor_channel_get(battery, SENSOR_CHAN_SOC, &charge_lvl);
+	if (rc) {
+		return;
+	}
+}
 
-void bas_init(void)
+static ssize_t charge_lvl_read(struct bt_conn *conn,
+			       const struct bt_gatt_attr *attr, void *buf,
+			       u16_t len, u16_t offset)
 {
-	bt_gatt_service_register(&bas_svc);
+	u8_t lvl8 = (u8_t)(charge_lvl.val1);
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &lvl8,
+				 sizeof(lvl8));
 }
 
-void bas_notify(void)
+void level_changed_callback(struct device *dev, struct sensor_trigger *trigger)
 {
-	if (!simulate_blvl) {
-		return;
+	ARG_UNUSED(dev);
+
+	if (trigger->type == SENSOR_TRIG_DATA_READY &&
+	    trigger->chan == SENSOR_CHAN_SOC) {
+		/* fetch battery level and send notifications, if enabled */
+		charge_lvl_fetch();
+		bas_notify();
 	}
+}
+
+static int bas_init(struct device *dev)
+{
+	int rc;
 
-	battery--;
+	ARG_UNUSED(dev);
+
+	battery = device_get_binding(CONFIG_BATTERY_SENSOR_DEV_NAME);
 	if (!battery) {
-		/* Software eco battery charger */
-		battery = 100;
+		return -EIO;
 	}
 
-	bt_gatt_notify(NULL, &attrs[1], &battery, sizeof(battery));
+	struct sensor_trigger trigger = {
+		.type = SENSOR_TRIG_DATA_READY,
+		.chan = SENSOR_CHAN_SOC
+	};
+
+	rc = sensor_trigger_set(battery, &trigger, level_changed_callback);
+	if (rc) {
+		return -EIO;
+	}
+
+	return bt_gatt_service_register(&bas);
 }
+
+int bas_notify(void)
+{
+	if (!notif_enabled) {
+		return -EINVAL;
+	}
+
+	u8_t lvl8 = (u8_t)(charge_lvl.val1);
```

Comment:
Id have the declaration in the beginning of the function, some c standard actually forbids declaration not in the beginning of the function.